builtin_array.go

```diff
@@ -185,19 +186,22 @@ func (r *Runtime) arrayproto_join(call FunctionCall) Value {
 	}
 
 	var buf valueStringBuilder
-
-	element0 := o.self.getIdx(valueInt(0), nil)
-	if element0 != nil && element0 != _undefined && element0 != _null {
-		buf.WriteString(element0.toString())
-	}
-
-	for i := 1; i < l; i++ {
-		buf.WriteString(sep)
+	for i := 0; i < l; i++ {
 		element := o.self.getIdx(valueInt(int64(i)), nil)
 		if element != nil && element != _undefined && element != _null {
-			buf.WriteString(element.toString())
+			obj := element.ToObject(r)
+			if _, ok := o.seenObjects[obj]; ok {
+				buf.WriteString(newStringValue("[Circular]"))
+			} else {
+				obj.AddSeenObjectMap(o.seenObjects)
+				buf.WriteString(obj.toString())
+			}
+		}
+		if i != l-1 {
+			buf.WriteString(sep)
 		}
```

Comment:
one of our main goals is to keep this code as close as possible to the main repo that we forked so that way future rebases will be easier. Can we either keep abstract out this new code into a func or at least move the `sep` logic first as the original repo has it?
```
for i := 0; i < l; i++ {
    if i != 0 {
                buf.WriteString(sep)
    }			
}
```